### PR TITLE
Watch CI/CD pipelines for completion and failure notifications

### LINKED ISSUE #266: Watch ADO pipelines and GitHub workflows for completion/failure notifications
## Summary

Allow users to "watch" ADO pipelines and GitHub Actions workflows to be notified of their completion, including job failures that occur while other jobs are still in progress.

## Motivation

Developers frequently kick off CI/CD pipelines and then context-switch. Currently there's no way to track pipeline/workflow status from within DevDocket. Users need to keep browser tabs open or rely on email notifications that may be delayed.

## Requirements

- Watch GitHub Actions workflow runs for completion/failure notifications
- Watch ADO pipeline runs for completion/failure notifications
- Notify on individual job failures even while the overall run is still in progress (early failure detection)
- Integrate into the DevDocket UI (exact placement TBD — needs design work)

## Design Considerations

This feature needs design discussion before implementation to determine:

1. **UI integration** — Where does pipeline/workflow status appear? Options include:
   - A new view in the activity bar (e.g., "Pipelines" or "CI/CD")
   - Inline status on work items that triggered the pipeline
   - A notification-only approach (no persistent UI)
   - Status bar indicators
2. **Watch lifecycle** — How does a user start/stop watching? Auto-detect from branch? Manual add?
3. **Notification style** — VS Code notifications, status bar changes, inline tree decorations, or a combination?
4. **Polling vs events** — GitHub and ADO have different APIs for run status; need to determine the right polling interval or webhook approach
5. **Provider architecture** — Should this be a new provider type, an extension of existing GitHub/ADO providers, or a separate concern?

## Labels

`enhancement`, `needs-design`

### PR DESCRIPTION
**Pull request by mthalman**
## Summary

Watch GitHub Actions workflow runs and Azure DevOps pipeline builds for status changes, firing VS Code notifications on completion or failure. Includes early individual job failure detection while runs are still in progress.

### Shared (packages/shared)
- **StatusWatcher\<T\>** utility class that detects status transitions across polling cycles. Newly observed items are silently tracked; only transitions from one known status to another produce change events. Designed for reuse across CI watchers and future PR tracking (#276).

### GitHub (packages/github)
- **GitHubActionsProvider**: New provider extending BaseGitHubProvider that polls \/repos/{owner}/{repo}/actions/runs\ for recent workflow runs. Tracks run statuses via StatusWatcher, fires VS Code notifications on completion (success/failure), and checks individual jobs within in-progress runs for early failure detection.
- Configuration: \devdocketGithub.watchWorkflowRuns\ (boolean, default: true)

### ADO (packages/ado)
- **AdoPipelineProvider**: New provider extending BaseProvider that polls the ADO Build API for recent builds. Uses the Build Timeline API for job-level failure detection within in-progress builds. Same notification pattern as the GitHub provider.
- Configuration: \devdocketAdo.watchPipelineRuns\ (boolean, default: true)

### Notification Behavior
- ✅ Success: \showInformationMessage\ with 'View Run'/'View Build' action button
- ❌ Failure: \showWarningMessage\ with action button to open the run/build URL
- ⚠️ Early job failure: \showWarningMessage\ for individual job failures within in-progress runs (deduplicated — each job failure notified only once per run)
- Cancelled runs are logged at info level, not notified

### Tests
- StatusWatcher: 11 tests covering new items, status changes, removals, and edge cases
- GitHubActionsProvider: 14 tests covering discovery, notifications, early failure detection, deduplication, and error handling
- AdoPipelineProvider: 13 tests covering the same patterns for ADO

Closes #266

### CODE REVIEW

File: packages/shared/src/statusWatcher.ts [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+   * @param current - Map of item IDs to their current status.
+   * @returns Array of status changes detected since the last update.
+   */
+  update(current: Map<string, TStatus>): StatusChange<TStatus>[] {
+    const changes: StatusChange<TStatus>[] = [];
+
+    for (const [id, status] of current) {
+      const previous = this.tracked.get(id);
+      if (previous !== undefined && previous !== status) {
+        changes.push({ id, previousStatus: previous, currentStatus: status });
```
[the remark this answers is not in the document]
Comment (mthalman) [reply packages/shared/src/statusWatcher.ts]: Fixed in 86fcbae

File: packages/ado/src/adoPipelineProvider.ts [hunk 2]
```diff
@@ -0,0 +1,335 @@
...
+
+    logger.info(`Fetched ${allBuilds.length} ADO pipeline builds`);
+
+    // Detect build-level status changes
+    const statusMap = new Map<string, string>();
+    for (const build of allBuilds) {
+      statusMap.set(String(build.id), compositeStatus(build));
+    }
+    const changes = this.buildWatcher.update(statusMap);
+    await this.notifyBuildChanges(changes, allBuilds);
```
[the remark this answers is not in the document]
Comment (mthalman) [reply packages/ado/src/adoPipelineProvider.ts:178]: Fixed in 5f63ec1

File: packages/ado/src/adoPipelineProvider.ts [hunk 3]
```diff
@@ -0,0 +1,335 @@
...
+    // Emit only non-completed builds as DiscoveredItems
+    const activeBuilds = allBuilds.filter(b => b.status !== 'completed');
+    const items: DiscoveredItem[] = activeBuilds.map(build => ({
+      externalId: `pipelines:${build.project.name}/builds/${build.id}`,
+      title: `${build.definition.name} #${build.buildNumber}`,
+      description: `${formatBranch(build.sourceBranch)} · ${build.status}`,
+      url: build._links.web.href,
+      group: build.project.name,
+      state: build.status,
+    }));
```
[the remark this answers is not in the document]
Comment (mthalman) [reply packages/ado/src/adoPipelineProvider.ts:203]: Fixed in 178df52

File: packages/github/src/githubActionsProvider.ts [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+      } else if (change.currentStatus === 'completed:failure') {
+        const result = await vscode.window.showWarningMessage(
+          `❌ ${label} failed (${run.repository.full_name}, ${branch})`,
+          'View Run',
+        );
+        if (result === 'View Run') {
+          void vscode.env.openExternal(vscode.Uri.parse(run.html_url));
+        }
+      } else if (change.currentStatus === 'completed:cancelled') {
+        logger.info(`Workflow run ${label} was cancelled (${run.repository.full_name}, ${branch})`);
```
[the remark this answers is not in the document]
Comment (mthalman) [reply packages/github/src/githubActionsProvider.ts]: Fixed in 47f72ee

File: packages/ado/src/adoPipelineProvider.ts [hunk 5]
```diff
@@ -0,0 +1,335 @@
...
+      } else if (change.currentStatus === 'completed:failed') {
+        const result = await vscode.window.showWarningMessage(
+          `❌ ${label} failed (${build.project.name}, ${branch})`,
+          'View Build',
+        );
+        if (result === 'View Build') {
+          void vscode.env.openExternal(vscode.Uri.parse(build._links.web.href));
+        }
+      } else if (change.currentStatus === 'completed:canceled') {
+        logger.info(`ADO build ${label} was cancelled (${build.project.name}, ${branch})`);
```
[the remark this answers is not in the document]
Comment (mthalman) [reply packages/ado/src/adoPipelineProvider.ts]: Fixed in 3c16446

[the remark this answers is not in the document]
Comment (mthalman) [reply packages/ado/src/adoPipelineProvider.ts]: Fixed in 3c16446 (addressed in same commit as comment 5)

File: packages/github/src/extension.ts [hunk 6]
```diff
@@ -72,14 +75,50 @@ export async function activate(_context: vscode.ExtensionContext): Promise<void>
...
+
+  configureActionsProvider(intervalSeconds);
+
+  _context.subscriptions.push(
+    vscode.workspace.onDidChangeConfiguration(e => {
+      if (e.affectsConfiguration('devdocketGithub.watchWorkflowRuns')) {
+        const ghConfig = vscode.workspace.getConfiguration('devdocketGithub');
+        const interval = validateRefreshInterval(
+          ghConfig.get<number>('refreshIntervalSeconds', 300), logger,
+        );
```
[the remark this answers is not in the document]
Comment (mthalman) [reply packages/github/src/extension.ts]: Fixed in 22c585f

File: packages/github/src/githubActionsProvider.ts [hunk 7]
```diff
@@ -0,0 +1,290 @@
...
+  private readonly runWatcher = new StatusWatcher<string>();
+  /** Tracks which job failures have already been notified per run to avoid duplicates. */
+  private readonly notifiedJobFailures = new Map<number, Set<string>>();
+
+  protected async fetchAndPublish(accessToken: string, _isUserTriggered: boolean): Promise<void> {
+    logger.info('Fetching GitHub Actions workflow runs...');
+    const repos = this.getConfiguredRepos();
+
+    if (repos.length === 0) {
+      logger.debug('No repos configured — skipping GitHub Actions fetch');
```
[the remark this answers is not in the document]
Comment (mthalman) [reply packages/github/src/githubActionsProvider.ts:81]: Fixed in b2e4aa1

File: packages/github/src/githubActionsProvider.ts [hunk 8]
```diff
@@ -0,0 +1,290 @@
...
+    const maxConcurrent = Math.min(3, runs.length);
+
+    const runWorker = async (): Promise<void> => {
+      while (nextIndex < runs.length) {
+        const idx = nextIndex++;
+        const run = runs[idx];
+        try {
+          const jobs = await this.fetchRunJobs(token, run);
+          const failedJobs = jobs.filter(j => j.status === 'completed' && j.conclusion === 'failure');
+
```
[the remark this answers is not in the document]
Comment (mthalman) [reply packages/github/src/githubActionsProvider.ts:245]: Fixed in 43753c8

File: packages/github/src/extension.ts [hunk 9]
```diff
@@ -72,14 +75,53 @@ export async function activate(_context: vscode.ExtensionContext): Promise<void>
...
+    vscode.workspace.onDidChangeConfiguration(e => {
+      if (
+        e.affectsConfiguration('devdocketGithub.watchWorkflowRuns') ||
+        e.affectsConfiguration('devdocketGithub.refreshIntervalSeconds')
+      ) {
+        const ghConfig = vscode.workspace.getConfiguration('devdocketGithub');
+        const interval = validateRefreshInterval(
+          ghConfig.get<number>('refreshIntervalSeconds', 300), logger,
+        );
+        configureActionsProvider(interval);
```
[the remark this answers is not in the document]
Comment (mthalman) [reply packages/github/src/extension.ts:117]: This is pre-existing behavior - the GitHub extension never had interval change handling for the issue/PR providers. Out of scope for this PR.

File: packages/ado/src/adoPipelineProvider.ts [hunk 10]
```diff
@@ -0,0 +1,358 @@
...
+          'View Build',
+        );
+        if (result === 'View Build') {
+          void vscode.env.openExternal(vscode.Uri.parse(build._links.web.href));
+        }
+      } else if (change.currentStatus === 'completed:canceled') {
+        logger.info(`ADO build ${label} was canceled (${build.project.name}, ${branch})`);
+      } else if (change.currentStatus.startsWith('completed:')) {
+        const buildResult = change.currentStatus.slice('completed:'.length);
+        const result = await vscode.window.showWarningMessage(
```
[the remark this answers is not in the document]
Comment (mthalman) [reply packages/ado/src/adoPipelineProvider.ts]: Fixed in db7ed51

File: packages/github/src/githubActionsProvider.ts [hunk 11]
```diff
@@ -0,0 +1,294 @@
...
+      state: run.status,
+    }));
+
+    this._onDidDiscoverItems.fire(items);
+
+    if (failures.length > 0) {
+      const msg = failures.length === 1
+        ? `Failed to fetch workflow runs from ${failures[0]}`
+        : `Failed to fetch workflow runs from ${failures.length} repositories`;
+      logger.warn(msg);
```
[the remark this answers is not in the document]
Comment (mthalman) [reply packages/github/src/githubActionsProvider.ts:156]: Fixed in d6af46e

File: .squad/agents/fenster/history.md [hunk 12]
```diff
@@ -49,12 +49,22 @@ DevDocket is a VS Code extension monorepo for managing work items from multiple
...
 #276 (auto-track authored PRs), #282 (provider state in editor), #281 (clickable title), #275 (History→Queue transitions),#273 (tree counts), #265 (auto-complete on close), #250 (group context), #249 (accept-to-focus, pre-shipped), #243 (version resurfacing), #240 (create from URL), #233 (provider health), #232 (clear history), #231 (sources icons), #230 (layout toggle), #229 (emoji removal), #227 (provider labels), #223 (dead code cleanup), #222 (responsive layout), #221 (contextual heading), #219 (source URL link), #217 (editor metadata), #216 (provider description), #215 (dynamic titles), #189 (dismissed fix), #178 (ADO filtering), #158 (markdown injection), #157 (API trust boundary), #156 (URL sanitization), #155 (URL scheme validation), #154 (crypto.randomUUID), #153 (JSON validation), #152 (path traversal fix), #12 (AI PR actions), bulk rename (WorkCenter→DevDocket)
```
[the remark this answers is not in the document]
Comment (mthalman) [reply .squad/agents/fenster/history.md]: Fixed in 82dc22b — merged the two duplicate lines into a single deduplicated entry.

File: packages/core/src/test/providerRegistry.test.ts [hunk 13]
```diff
@@ -1384,12 +1384,15 @@ describe('ProviderRegistry', () => {
       const listener = vi.fn();
       registry.onDidChangeProviderHealth(listener);
 
-      // Guarantee a different timestamp on next refresh
-      const spy = vi.spyOn(Date, 'now').mockReturnValue(Date.now() + 1000);
+      // Guarantee a different timestamp on next refresh.
+      // vi.useFakeTimers + setSystemTime properly mocks new Date(),
```
[the remark this answers is not in the document]
Comment (mthalman) [reply packages/core/src/test/providerRegistry.test.ts]: Already applied — the comment on line 1388 already reads `vi.useFakeTimers({ now })` matching the actual code.

File: packages/ado/src/test/extension.test.ts [hunk 14]
```diff
@@ -125,19 +125,21 @@ describe('extension activation', () => {
...
-    expect(mockApi.registerProvider).toHaveBeenCalledTimes(2);
+    expect(mockApi.registerProvider).toHaveBeenCalledTimes(3);
   });
 
-  it('registers two providers when organization is configured', async () => {
+  it('registers three providers when organization is configured', async () => {
     await activate(mockContext);
 
-    expect(mockApi.registerProvider).toHaveBeenCalledTimes(2);
+    expect(mockApi.registerProvider).toHaveBeenCalledTimes(3);
```
[the remark this answers is not in the document]
Comment (mthalman) [reply packages/ado/src/test/extension.test.ts]: Already addressed - the test names were updated to 'when ADO projects are configured' and 'when no ADO projects are configured' in an earlier commit.

[the remark this answers is not in the document]
Comment (mthalman) [reply packages/ado/src/test/extension.test.ts]: Fixed in b9049f0 — renamed test descriptions to reference ADO projects config instead of organization.

File: packages/github/src/test/githubActionsProvider.test.ts [hunk 15]
```diff
@@ -0,0 +1,393 @@
...
+    const runs = [
+      createMockRun(1, 'CI', 'in_progress', null, 'owner/repo', 'feature-branch'),
+      createMockRun(2, 'Deploy', 'queued', null, 'owner/repo', 'main'),
+      createMockRun(3, 'Tests', 'completed', 'success', 'owner/repo', 'main'),
+    ];
+
+    mockFetch.mockResolvedValueOnce({
+      ok: true,
+      json: async () => ({ workflow_runs: runs }),
+    });
```
[the remark this answers is not in the document]
Comment (mthalman) [reply packages/github/src/test/githubActionsProvider.test.ts:122]: Fixed in 384ed9c — added mock for the jobs endpoint so the test stays on the happy path.
